### PR TITLE
feat(rust, python): Add `str.extract_groups`

### LINKED ISSUE #3775: col.str.extract() multiple groups at once
#### Describe your feature request

I often find myself dealing with having to extract features of interest from a string. Something like, let's say, this:

```py
import polars as pl

iso_specs = pl.DataFrame({"iso_code": ["ISO 1000:1992", "ISO 80000-1:2009", "ISO/IEC 18181-1:2022", "ISO/IEC/IEEE 29148:2018", "ISO/IEC 9075-1:2016", "ISO 9660:1988"]})
```

With a quick (and completely unreadable) regex, I can extract various features from those strings:

image

If I need to turn these features into polars columns, however, while that's completely doable, it's certainly not... great.

```py
import polars as pl

UNREADABLE_REGEX = r"(^ISO(?:/[A-Z]+)*)\s(\d+)(?:-(\d+))?(?::(\d{4}))"

iso_specs = pl.DataFrame({"iso_code": ["ISO 1000:1992", "ISO 80000-1:2009", "ISO/IEC 18181-1:2022", "ISO/IEC/IEEE 29148:2018", "ISO/IEC 9075-1:2016", "ISO 9660:1988"]})

iso_specs.select([
    pl.col("iso_code"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 1).alias("authority"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 2).alias("spec_num"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 3).alias("part_num"),
    pl.col("iso_code").str.extract(UNREADABLE_REGEX, 4).alias("revision_year"),
])
```

Given all these columns come from the same regex, they could all be extracted in one go, but I get the impression that the optimizer may not catch that and will just run the regex 4 times. It's also a bit long to write.

Ideally, I would like something like this:

```py
iso_specs.select([
    pl.col("iso_code"),
    pl.col("iso_code").str.extract_groups(UNREADABLE_REGEX, [1,2,3,4]),
])
```

...but then this runs into problems with the names of multicolumn expressions, since `alias()` only works on single columns and `map_alias()` wouldn't be able to make the difference between each.

------

Since #2848 `split_exact()` [has a workaround for the column name issue](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.internals.expr.string.ExprStringNameSpace.split_exact.html) it returns a struct that can be unnested, though I'm not personally fond of the ergonomics.

```py
csv_data = csv_data.with_columns([
    pl.col("timestamp").str.strptime(pl.Datetime, fmt="%Y-%m-%d %H:%M:%S", exact=True, strict=True),
    
    pl.col("gps_coords").str.split_exact(";", 1).struct.rename_fields(["lat", "long"]),
]).unnest("gps_coords")

# I need to cut the expression in half so I can unnest the columns first
csv_data = csv_data.with_columns([
    pl.col("lat").cast(pl.Float64),
    pl.col("long").cast(pl.Float64),
])
```

EDIT: Alternatively, without unnesting:

```py
df = pl.DataFrame(
    {
		"package": ["polars","pandas","python"],
		"version": ["0.14.18","1.4.4","3.10.7"],
    }
)

df = df.with_columns([
	pl.col("version").str.split_exact(".", n=2).struct.rename_fields(
 		["semver_major","semver_minor","semver_revision"]
 	).alias("semver_struct")
]).with_columns([
	pl.col("semver_struct").struct.field("semver_major"),
	pl.col("semver_struct").struct.field("semver_minor"),
	pl.col("semver_struct").struct.field("semver_revision"),
])
```

### PR DESCRIPTION
**Pull request by cmdlineluser**
Attempt to resolve #3775 (also resolve https://github.com/pola-rs/polars/issues/7990)

I saw this issue and was curious if I could get it working.

I've managed to get a working version, but not sure on the implementation.

Not entirely sure on the most optimal way to create the `StructChunked`.

`.captures` is just a "working title" from the underlying rust function, a better name can be chosen.

**Update:** now known as `.extract_groups()`
```python
df = pl.DataFrame({"col": ["abc", "de", None, "ghi"]})

df.with_columns(
   pl.all().str.extract_groups(r"\A(?<one>.)(.)(?<three>.)\z")
)

# shape: (4, 1)
# ┌──────────────────┐
# │ col              │
# │ ---              │
# │ struct[3]        │
# ╞══════════════════╡
# │ {"a","b","c"}    │
# │ {null,null,null} │
# │ {null,null,null} │
# │ {"g","h","i"}    │
# └──────────────────┘

df.with_columns(
   pl.all().str.extract_groups(r"\A(?<one>.)(.)(?<three>.)\z")
).unnest("col")

# shape: (4, 3)
# ┌──────┬──────┬───────┐
# │ one  ┆ 2    ┆ three │
# │ ---  ┆ ---  ┆ ---   │
# │ str  ┆ str  ┆ str   │
# ╞══════╪══════╪═══════╡
# │ a    ┆ b    ┆ c     │
# │ null ┆ null ┆ null  │
# │ null ┆ null ┆ null  │
# │ g    ┆ h    ┆ i     │
# └──────┴──────┴───────┘
```

### CODE REVIEW

**Comment by avimallu** [comment]
This will be a great feature!

May I suggest calling it `str.extract_multiple`?

**Comment by cmdlineluser** [comment]
Oops, I probably should have specified that I wasn't voting for it to be called `.captures` - I just needed to name it something.

I've edited that in, thanks @avimallu

**Comment by ion-elgreco** [comment]
> This will be a great feature!
> 
> May I suggest calling it `str.extract_multiple`?

How about `.extract_groups`, then it's more clear you can use it for capture groups. Initially, I thought the `.extract_all` would capture all groups, but the `all` was misleading.

**Comment by alexander-beedie** [comment]
Nice one :)  As for naming, `extract_captures` or `extract_groups` would both be reasonable, and will naturally group the functions together for autocomplete/etc.

**Comment by ion-elgreco** [comment]
> Nice one :)  As for naming, `extract_captures` or `extract_groups` would both be reasonable (maybe the latter has a small edge?), and will naturally group the functions together for autocomplete/etc.

`Extract_groups` would be slightly more clear since `.extract` uses group_index. Then one could understand from the name its .extract but with all groups at once.

**Comment by alexander-beedie** [comment]
> `Extract_groups` would be slightly more clear since `.extract` uses group_index. Then one could understand from the name its .extract but with all groups at once.

Works for me... 👍

**Comment by sm-Fifteen** [comment]
> > Nice one :)  As for naming, `extract_captures` or `extract_groups` would both be reasonable (maybe the latter has a small edge?), and will naturally group the functions together for autocomplete/etc.
> 
> `Extract_groups` would be slightly more clear since `.extract` uses group_index. Then one could understand from the name its .extract but with all groups at once.

From the examples, it looks like match 0 is always the full pattern, so `captures` might be a bit less misleading in that regard.

**Comment by cmdlineluser** [comment]
Yeah, I was actually just about to ask if the 0th group should be returned at all from this function?

File: py-polars/polars/series/string.py [hunk 1]
```diff
@@ -770,6 +770,42 @@ def extract_all(self, pattern: str | Series) -> Series:
...
+        --------
+        >>> s = pl.Series(
+        ...     name="url",
+        ...     values=[
+        ...         "http://vote.com/ballon_dor?candidate=messi&ref=python",
+        ...         "http://vote.com/ballon_dor?candidate=weghorst&ref=polars",
+        ...         "http://vote.com/ballon_dor?error=404&ref=rust",
+        ...     ],
+        ... )
+        >>> s.str.extract_groups(r"candidate=(?<candidate>\w+)&ref=(?<ref>\w+)")
```
[the remark this answers is not in the document]
Comment (cmdlineluser) [reply py-polars/polars/series/string.py:798]: > but numbered fields should still be mentioned

Good point, thanks.

> naming their third capture group "2"

Named captures must be a "valid identifier", you cannot create a group named "2".

[the remark this answers is not in the document]
Comment (sm-Fifteen) [reply py-polars/polars/series/string.py:798]: > > naming their third capture group "2"
> 
> Named captures must be a "valid identifier", you cannot create a group named "2".

Ah, that's reassuring. I had missed the part in the Rust doc that said "Names must start with either an _ or an alphabetic codepoint".

**Comment by mcrumiller** [comment]
This will also close #7990.

File: py-polars/tests/unit/namespaces/test_string.py [hunk 2]
```diff
@@ -537,6 +537,39 @@ def test_extract_all_many() -> None:
     assert df["foo"].str.extract_all(df["re"]).to_list() == [["a"], ["bc"], ["abc"]]
 
 
+def test_extract_groups() -> None:
```
Comment (stinodego) [inline py-polars/tests/unit/namespaces/test_string.py:540]: This is a nice complex test case. Could we also add a very simple test for both named groups and numeric groups?

**Review by stinodego — COMMENTED** [review]
This looks really good :ok_hand:

I fixed some Sphinx RST.

Would like another test - maybe you can think of some edge cases yourself and test for those? What if there are no capturing groups? etc.

Maybe @ritchie46 or @orlp can judge whether the Rust code is optimal here.

File: py-polars/tests/unit/namespaces/test_string.py (hunk 2 above)
Comment (cmdlineluser) [reply py-polars/tests/unit/namespaces/test_string.py:540]: Yeah, good point - thanks. More testing is needed.

No groups was actually not handled correctly:
```python
df.select(pl.col("iso_code").str.extract_groups(""))

# shape: (1, 1)
# ┌───────────┐
# │ iso_code  │
# │ ---       │
# │ struct[1] │
# ╞═══════════╡
# │ {null}    │
# └───────────┘
```
I've just updated it to cast the original column to null
```python
df.select(pl.col("iso_code").str.extract_groups(""))

# shape: (2, 1)
# ┌───────────┐
# │ iso_code  │
# │ ---       │
# │ struct[1] │
# ╞═══════════╡
# │ {null}    │
# │ {null}    │
# └───────────┘

df.select(pl.col("iso_code").str.extract_groups("")).schema
# {'iso_code': Struct([Field('iso_code', Null)])}
```
Not sure what the "correct" field name should be in this case.

**Comment by ritchie46** [comment]
Thanks for the PR @cmdlineluser. Looks very good. 

I added feature gates as I don't want this functionality to compile on the rust side if users don't need it. 

Other than that I also moved the implementation to the array level as we shouldn't iterate over chunkedarrays.

**Pull request merged**